File: ai-escape-room-api/src/schemas/levels.py
from pydantic import BaseModel, Field
from typing import Optional
from typing import Literal
import uuid
from datetime import datetime

class RateLevelRequest(BaseModel):
    rate: int = Field(min=1, max=5)

class LevelListItem(BaseModel):
    id: uuid.UUID
    title: str
    story: str
    difficulty: int
    generated_at: datetime
    rating: Optional[float] = None
    favorite_count: int = 0
    total_tokens: int = 0
    total_minutes: float = 0.0
    repair_count: int = 0

class LevelListQuery(BaseModel):
    title: str | None = None
    story: str | None = None
    rating_gte: Literal[3, 4] | None = None
    difficulty: Literal[1, 2, 3, 4, 5] | None = None
    favorites_only: bool = False

File: ai-escape-room/src/api/types/levels.ts
import { SpriteStyleEnum } from "../../shared/enums"

export type RatingFilter = "3" | "4" | ""
export type DifficultyFilter = "1" | "2" | "3" | "4" | "5" | ""

export type LevelListItem = {
  id: string
  title: string
  story: string
  difficulty: number
  generated_at: string
  rating: number | null
  favorite_count: number
  total_tokens: number
  total_minutes: number
  repair_count: number
  avg_completion_minutes: number | null
}

export type PagedResponse<T> = {
  items: T[]
  page: number
  page_size: number
  total: number
  total_pages: number
}

export type ListLevelsQuery = {
  page: number
  pageSize: number
  title?: string
  story?: string
  ratingGte?: RatingFilter
  difficulty?: DifficultyFilter
  favoritesOnly?: boolean
}

export interface EstimateTokensRequest {
  difficulty: number
}

export interface EstimateTokensResponse {
  estimated_tokens: number
  estimated_minutes: number
  current_balance: number
  sufficient: boolean
}

export interface GenerateLevelRequest {
  difficulty: number
  sprite_style: SpriteStyleEnum
  story: string
}

export interface GenerateLevelResponse {
  level: any
  tokens: {
    generation_tokens: number
    generation_minutes: number
    validation_tokens: number
    validation_minutes: number
    repair_tokens: number
    repair_minutes: number
    sprite_tokens: number
    sprite_minutes: number
    total_tokens: number
    total_minutes: number
    repair_count: number
  }
}

export interface LevelCompletionRequest {
  completion_minutes: number
}
